### PR TITLE
[teamup-rocketchat] Added support for custom config file and log path location via cli parameters

### PR DESCRIPTION
**Pull request by roshanlc**
This PR addresses issues #6 and #7 since both issues are interrelated. 

Changes Added :
- Support for custom config file path via `--config` parameter
- Support for custom log path via `--logpath` parameter
- Shows usage  and help screen with `-h` and `--help` parameter

### CODE REVIEW

File: teamup-rocketchat-bot/README.md [hunk 1]
```diff
@@ -5,7 +5,49 @@
 ### Currently supports
 
 - Reminding upcoming meeting events
+### Configuration
+- The configuration file should have the `yml` extension and the following properties.
+```bash
+# Url of rocket chat server
+URL: https:rocket.chat
```
Comment (individual-it) [inline teamup-rocketchat-bot/README.md:12]:
```suggestion
URL: https://rocket.chat
```

File: teamup-rocketchat-bot/main.go [hunk 2]
```diff
@@ -229,7 +227,7 @@ func checkForMeetings(config *Configuration, chatClient *gorocket.Client) {
 		toSendMsgs := []string{}
 		for _, event := range futureEvents {
 			diff := timeDiffWithNow(event.StartDt)
-			if diff > 10 && diff < 21 {
+			if diff > 1 && diff < 21 {
```
Comment (individual-it) [inline teamup-rocketchat-bot/main.go:230]: why this change and what does it has to do with the config file/cli parameters?

**Review by individual-it — CHANGES_REQUESTED** [review]
some small initial comments, I still need to try it

File: teamup-rocketchat-bot/main.go [hunk 3]
```diff
@@ -6,25 +6,36 @@ import (
...
-	"github.com/go-yaml/yaml"
 )
 
-const fileName = "config.yml"
+const defaultConfigFileName = "config.yml"
 
 var ErrLogin = errors.New("the login was not successfull most likely due to invalid credentails")
 
-const logFile = "latest.log"
+const defaultLogFilePath = "/tmp"
```
Comment (individual-it) [inline teamup-rocketchat-bot/main.go:25]: the usual path for log files would be `/var/log`

File: teamup-rocketchat-bot/configuration.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	LogFileName  string `yaml:"LOG_FILE_NAME"`
+}
+
+// Prints beatiful
+func (config Configuration) String() string {
+	return fmt.Sprintf(
+		"URL:%s\nUsername:%v\nPassword:%s\nUseSSL:%v\nRoom:%s\nMeetingsCode:%s\nTeamupToken:%s\nRepeatIn:%d\nLogPath:%s\nLogFileName:%s\n",
+		config.URL,
+		config.Username,
+		config.Password,
```
Comment (individual-it) [inline teamup-rocketchat-bot/configuration.go:33]: should also be hidden

File: teamup-rocketchat-bot/main.go [hunk 5]
```diff
@@ -255,29 +253,104 @@ func checkForMeetings(config *Configuration, chatClient *gorocket.Client) {
...
+	var defaultPath string
+	// has CustomPath been used
+	var isCustomPath bool = false
+	if len(customLogPath) == 0 {
+		defaultPath = defaultLogFilePath + "/" + defaultLogFileName
+	} else {
+		defaultPath = path.Clean(customLogPath) + "/" + defaultLogFileName
+		isCustomPath = true
+	}
+	defaultLog, err := os.OpenFile(defaultPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644) // 0644 = user can read and write, other and groups can read
```
Comment (individual-it) [inline teamup-rocketchat-bot/main.go:304]: if there are not enough permissions to create the file the bot crashes ugly
```
TEAMUP-ROCKETCHAT-BOT 2023/01/11 09:05:30 Looking for configuration at "/var/www/rocket.chat/teamup/config.yml". Setting directory specified by configuration for log output.
TEAMUP-ROCKETCHAT-BOT 2023/01/11 09:05:30 open /tmp/teamup-rocket-chat.log: permission denied
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x529229]

goroutine 1 [running]:
log.(*Logger).Output(0x0, 0x1?, {0xc000128200, 0x33})
	/usr/local/go/src/log/log.go:179 +0x89
log.(*Logger).Fatal(0xc0001281c0?, {0xc000125f00?, 0x441?, 0x7c41f0?})
	/usr/local/go/src/log/log.go:231 +0x45
main.main()
	/tmp/hubot/teamup-rocketchat-bot/main.go:307 +0x158
```

also I would suggest to make the permissions 0640, others (anyone on the system) does not need to be able to read the logfile

**Review by individual-it — CHANGES_REQUESTED** [review]
this version is currently running for us as company and seems to work generally good.

you could also change the README for autostart it would now work simpler:
`ExecStart=<full-path-of-bot>/teamup-rocketchat-bot --config <full-path-of-config-file>`

File: teamup-rocketchat-bot/main.go [hunk 6]
```diff
@@ -6,25 +6,36 @@ import (
...
-const fileName = "config.yml"
+const defaultConfigFileName = "config.yml"
 
 var ErrLogin = errors.New("the login was not successfull most likely due to invalid credentails")
 
-const logFile = "latest.log"
+const defaultLogFilePath = "/tmp"
+const defaultLogFileName = "teamup-rocket-chat.log"
+
+const eventsTrackerFile = "/tmp/events_tracker.json"
```
[the remark this answers is not in the document]
Comment (roshanlc) [reply teamup-rocketchat-bot/main.go:28]: That would require the bot to be run with root privileges.

File: teamup-rocketchat-bot/main.go (hunk 5 above)
Comment (roshanlc) [reply teamup-rocketchat-bot/main.go:304]: I have  made an issue at [https://github.com/JankariTech/hubot/issues/15](https://github.com/JankariTech/hubot/issues/15) . Let's not add more features in this PR.

**Comment by individual-it** [comment]
also had a crash today, maybe because the rocket chat server was not reachable
image

File: teamup-rocketchat-bot/main.go [hunk 7]
```diff
@@ -255,37 +256,121 @@ func checkForMeetings(config *Configuration, chatClient *gorocket.Client) {
...
+	} else {
+		customPath = path.Clean(config.LogPath) + "/" + config.LogFileName
+	}
+	customLog, err := os.OpenFile(customPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644) // 0644 = user can read and write, other and groups can read
+
+	if err != nil {
+		log.Println(err.Error()) // Also print to stdout
+		logger.Fatal(err.Error())
 	}
+	log.Printf("The file at %q has been beet set for log output.\n", customPath)
```
Comment (individual-it) [inline teamup-rocketchat-bot/main.go:358]:
```suggestion
	log.Printf("The file at %q has been set for log output.\n", customPath)
```

File: teamup-rocketchat-bot/main.go [hunk 8]
```diff
@@ -255,37 +256,121 @@ func checkForMeetings(config *Configuration, chatClient *gorocket.Client) {
...
+	var defaultPath string
+	// has CustomPath been used
+	var isCustomPath bool = false
+	if len(customLogPath) == 0 {
+		defaultPath = defaultLogFilePath + "/" + defaultLogFileName
+	} else {
+		defaultPath = path.Clean(customLogPath) + "/" + defaultLogFileName
+		isCustomPath = true
+	}
+	defaultLog, err := os.OpenFile(defaultPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644) // 0644 = user can read and write, other and groups can read
```
Comment (individual-it) [inline teamup-rocketchat-bot/main.go:317]: no need to be so open, `0640` should be enough, same for the file creation further down, or even better have just one `OpenFile` call

**Review by individual-it — CHANGES_REQUESTED** [review]

**Review by individual-it — APPROVED** [review]
